### DISCUSSION TITLE
Graphviz-dot diagrams malformatted in html output

### DISCUSSION THREAD

**Discussion by Gewerd-Strauss** [quarto-dev/quarto-cli#6153] [Q&A] [answered]
### Description

Take the following example manuscript:

````qmd
---
key: X
---
```{dot}

digraph G {
	graph [fontsize=10 fontname="Verdana" compound=true];
	node [shape=record fontsize=6 fontname="Verdana"];
	subgraph cluster_outer {
    subgraph cluster_Y {
        label="directions"
             "P6 (bottom left, front)" -> "P1 (top, back)"
         -> "P2 (top, front)"
         -> "P3 (bottom right, back)"
         -> "P4 (bottom right, front)"
         -> "P5 (bottom left, back)"
         -> "P6 (bottom left, front)"
    }
    subgraph cluster_bl {
        node [style=radial]
        "P5 (bottom left, back)" [color=greenyellow]
        "P6 (bottom left, front)" [color=green]
        label="Bottom left shelf"
        color=red
    }
    subgraph cluster_br {
        node [style=radial]
        "P3 (bottom right, back)" [color=gold]
        "P4 (bottom right, front)" [color=yellow]
        label="Bottom right shelf"
        color=yellow
    }
    subgraph cluster_t {
        node [style=radial]
        "P1 (top, back)" [color=red]
        "P2 (top, front)" [color=orange]
        label="Top shelf"
        color=green
    }
    subgraph cluster_steps {
        label="Steps"
        node [style=filled];
     "1. Entferne Gruppe auf Position P5" [color=greenyellow]
     "2. Entferne Gruppe auf Position P6" [color=green]
     "3. Schiebe Gruppe P5 auf Position P6" [color=greenyellow]
     "4. Schiebe Gruppe P4 auf Position P5" [color=yellow]
     "5. Schiebe Gruppe P3 auf Position P4" [color=gold]
     "6. Schiebe Gruppe P2 auf Position P3" [color=orange]
     "7. Schiebe Gruppe P1 auf Position P2" [color=red]
        node [style=filled];
        "1. Entferne Gruppe auf Position P5" 
        -> "2. Entferne Gruppe auf Position P6"
        -> "3. Schiebe Gruppe P5 auf Position P6"
        -> "4. Schiebe Gruppe P4 auf Position P5"
        -> "5. Schiebe Gruppe P3 auf Position P4"
        -> "6. Schiebe Gruppe P2 auf Position P3"
        -> "7. Schiebe Gruppe P1 auf Position P2"
    
        color=black;
        hidden=true
    }
    subgraph cluster_1 {
        label = "Gruppen";
        node [style=filled];
        "G1" [color=red]
        "G2" [color=orange]
        "G3" [color=gold]
        "G4" [color=yellow]
        "G5" [color=greenyellow]
        "G6" [color=green]
        
        node [style=filled]
        {"G1"
        ->"G2" 
        ->"G3"
        ->"G4"
        ->"G5"
        ->"G6"[style=invis] {rank=0.1} }
        color=black;
        hidden=true;
        }
	}
}

```


Lorem ipsum

````

When building online, this diagram gets correctly rendered (see [here](https://dreampuf.github.io/GraphvizOnline/#%0Adigraph%20G%20%7B%0A%09graph%20%5Bfontsize%3D10%20fontname%3D%22Verdana%22%20compound%3Dtrue%5D%3B%0A%09node%20%5Bshape%3Drecord%20fontsize%3D6%20fontname%3D%22Verdana%22%5D%3B%0A%09subgraph%20cluster_outer%20%7B%0A%20%20%20%20subgraph%20cluster_Y%20%7B%0A%20%20%20%20%20%20%20%20label%3D%22directions%22%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%22P6%20(bottom%20left%2C%20front)%22%20-%3E%20%22P1%20(top%2C%20back)%22%0A%20%20%20%20%20%20%20%20%20-%3E%20%22P2%20(top%2C%20front)%22%0A%20%20%20%20%20%20%20%20%20-%3E%20%22P3%20(bottom%20right%2C%20back)%22%0A%20%20%20%20%20%20%20%20%20-%3E%20%22P4%20(bottom%20right%2C%20front)%22%0A%20%20%20%20%20%20%20%20%20-%3E%20%22P5%20(bottom%20left%2C%20back)%22%0A%20%20%20%20%20%20%20%20%20-%3E%20%22P6%20(bottom%20left%2C%20front)%22%0A%20%20%20%20%7D%0A%20%20%20%20subgraph%20cluster_bl%20%7B%0A%20%20%20%20%20%20%20%20node%20%5Bstyle%3Dradial%5D%0A%20%20%20%20%20%20%20%20%22P5%20(bottom%20left%2C%20back)%22%20%5Bcolor%3Dgreenyellow%5D%0A%20%20%20%20%20%20%20%20%22P6%20(bottom%20left%2C%20front)%22%20%5Bcolor%3Dgreen%5D%0A%20%20%20%20%20%20%20%20label%3D%22Bottom%20left%20shelf%22%0A%20%20%20%20%20%20%20%20color%3Dred%0A%20%20%20%20%7D%0A%20%20%20%20subgraph%20cluster_br%20%7B%0A%20%20%20%20%20%20%20%20node%20%5Bstyle%3Dradial%5D%0A%20%20%20%20%20%20%20%20%22P3%20(bottom%20right%2C%20back)%22%20%5Bcolor%3Dgold%5D%0A%20%20%20%20%20%20%20%20%22P4%20(bottom%20right%2C%20front)%22%20%5Bcolor%3Dyellow%5D%0A%20%20%20%20%20%20%20%20label%3D%22Bottom%20right%20shelf%22%0A%20%20%20%20%20%20%20%20color%3Dyellow%0A%20%20%20%20%7D%0A%20%20%20%20subgraph%20cluster_t%20%7B%0A%20%20%20%20%20%20%20%20node%20%5Bstyle%3Dradial%5D%0A%20%20%20%20%20%20%20%20%22P1%20(top%2C%20back)%22%20%5Bcolor%3Dred%5D%0A%20%20%20%20%20%20%20%20%22P2%20(top%2C%20front)%22%20%5Bcolor%3Dorange%5D%0A%20%20%20%20%20%20%20%20label%3D%22Top%20shelf%22%0A%20%20%20%20%20%20%20%20color%3Dgreen%0A%20%20%20%20%7D%0A%20%20%20%20subgraph%20cluster_steps%20%7B%0A%20%20%20%20%20%20%20%20label%3D%22Steps%22%0A%20%20%20%20%20%20%20%20node%20%5Bstyle%3Dfilled%5D%3B%0A%20%20%20%20%20%221.%20Entferne%20Gruppe%20auf%20Position%20P5%22%20%5Bcolor%3Dgreenyellow%5D%0A%20%20%20%20%20%222.%20Entferne%20Gruppe%20auf%20Position%20P6%22%20%5Bcolor%3Dgreen%5D%0A%20%20%20%20%20%223.%20Schiebe%20Gruppe%20P5%20auf%20Position%20P6%22%20%5Bcolor%3Dgreenyellow%5D%0A%20%20%20%20%20%224.%20Schiebe%20Gruppe%20P4%20auf%20Position%20P5%22%20%5Bcolor%3Dyellow%5D%0A%20%20%20%20%20%225.%20Schiebe%20Gruppe%20P3%20auf%20Position%20P4%22%20%5Bcolor%3Dgold%5D%0A%20%20%20%20%20%226.%20Schiebe%20Gruppe%20P2%20auf%20Position%20P3%22%20%5Bcolor%3Dorange%5D%0A%20%20%20%20%20%227.%20Schiebe%20Gruppe%20P1%20auf%20Position%20P2%22%20%5Bcolor%3Dred%5D%0A%20%20%20%20%20%20%20%20node%20%5Bstyle%3Dfilled%5D%3B%0A%20%20%20%20%20%20%20%20%221.%20Entferne%20Gruppe%20auf%20Position%20P5%22%20%0A%20%20%20%20%20%20%20%20-%3E%20%222.%20Entferne%20Gruppe%20auf%20Position%20P6%22%0A%20%20%20%20%20%20%20%20-%3E%20%223.%20Schiebe%20Gruppe%20P5%20auf%20Position%20P6%22%0A%20%20%20%20%20%20%20%20-%3E%20%224.%20Schiebe%20Gruppe%20P4%20auf%20Position%20P5%22%0A%20%20%20%20%20%20%20%20-%3E%20%225.%20Schiebe%20Gruppe%20P3%20auf%20Position%20P4%22%0A%20%20%20%20%20%20%20%20-%3E%20%226.%20Schiebe%20Gruppe%20P2%20auf%20Position%20P3%22%0A%20%20%20%20%20%20%20%20-%3E%20%227.%20Schiebe%20Gruppe%20P1%20auf%20Position%20P2%22%0A%20%20%20%20%0A%20%20%20%20%20%20%20%20color%3Dblack%3B%0A%20%20%20%20%20%20%20%20hidden%3Dtrue%0A%20%20%20%20%7D%0A%20%20%20%20subgraph%20cluster_1%20%7B%0A%20%20%20%20%20%20%20%20label%20%3D%20%22Gruppen%22%3B%0A%20%20%20%20%20%20%20%20node%20%5Bstyle%3Dfilled%5D%3B%0A%20%20%20%20%20%20%20%20%22G1%22%20%5Bcolor%3Dred%5D%0A%20%20%20%20%20%20%20%20%22G2%22%20%5Bcolor%3Dorange%5D%0A%20%20%20%20%20%20%20%20%22G3%22%20%5Bcolor%3Dgold%5D%0A%20%20%20%20%20%20%20%20%22G4%22%20%5Bcolor%3Dyellow%5D%0A%20%20%20%20%20%20%20%20%22G5%22%20%5Bcolor%3Dgreenyellow%5D%0A%20%20%20%20%20%20%20%20%22G6%22%20%5Bcolor%3Dgreen%5D%0A%20%20%20%20%20%20%20%20%0A%20%20%20%20%20%20%20%20node%20%5Bstyle%3Dfilled%5D%0A%20%20%20%20%20%20%20%20%7B%22G1%22%0A%20%20%20%20%20%20%20%20-%3E%22G2%22%20%0A%20%20%20%20%20%20%20%20-%3E%22G3%22%0A%20%20%20%20%20%20%20%20-%3E%22G4%22%0A%20%20%20%20%20%20%20%20-%3E%22G5%22%0A%20%20%20%20%20%20%20%20-%3E%22G6%22%5Bstyle%3Dinvis%5D%20%7Brank%3D0.1%7D%20%7D%0A%20%20%20%20%20%20%20%20color%3Dblack%3B%0A%20%20%20%20%20%20%20%20hidden%3Dtrue%3B%0A%20%20%20%20%20%20%20%20%7D%0A%09%7D%0A%7D%0A) and [here](http://magjac.com/graphviz-visual-editor/?dot=%0Adigraph%20G%20%7B%0A%09graph%20%5Bfontsize%3D10%20fontname%3D%22Verdana%22%20compound%3Dtrue%5D%3B%0A%09node%20%5Bshape%3Drecord%20fontsize%3D6%20fontname%3D%22Verdana%22%5D%3B%0A%09subgraph%20cluster_outer%20%7B%0A%20%20%20%20subgraph%20cluster_Y%20%7B%0A%20%20%20%20%20%20%20%20label%3D%22directions%22%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%22P6%20%28bottom%20left%2C%20front%29%22%20-%3E%20%22P1%20%28top%2C%20back%29%22%0A%20%20%20%20%20%20%20%20%20-%3E%20%22P2%20%28top%2C%20front%29%22%0A%20%20%20%20%20%20%20%20%20-%3E%20%22P3%20%28bottom%20right%2C%20back%29%22%0A%20%20%20%20%20%20%20%20%20-%3E%20%22P4%20%28bottom%20right%2C%20front%29%22%0A%20%20%20%20%20%20%20%20%20-%3E%20%22P5%20%28bottom%20left%2C%20back%29%22%0A%20%20%20%20%20%20%20%20%20-%3E%20%22P6%20%28bottom%20left%2C%20front%29%22%0A%20%20%20%20%7D%0A%20%20%20%20subgraph%20cluster_bl%20%7B%0A%20%20%20%20%20%20%20%20node%20%5Bstyle%3Dradial%5D%0A%20%20%20%20%20%20%20%20%22P5%20%28bottom%20left%2C%20back%29%22%20%5Bcolor%3Dgreenyellow%5D%0A%20%20%20%20%20%20%20%20%22P6%20%28bottom%20left%2C%20front%29%22%20%5Bcolor%3Dgreen%5D%0A%20%20%20%20%20%20%20%20label%3D%22Bottom%20left%20shelf%22%0A%20%20%20%20%20%20%20%20color%3Dred%0A%20%20%20%20%7D%0A%20%20%20%20subgraph%20cluster_br%20%7B%0A%20%20%20%20%20%20%20%20node%20%5Bstyle%3Dradial%5D%0A%20%20%20%20%20%20%20%20%22P3%20%28bottom%20right%2C%20back%29%22%20%5Bcolor%3Dgold%5D%0A%20%20%20%20%20%20%20%20%22P4%20%28bottom%20right%2C%20front%29%22%20%5Bcolor%3Dyellow%5D%0A%20%20%20%20%20%20%20%20label%3D%22Bottom%20right%20shelf%22%0A%20%20%20%20%20%20%20%20color%3Dyellow%0A%20%20%20%20%7D%0A%20%20%20%20subgraph%20cluster_t%20%7B%0A%20%20%20%20%20%20%20%20node%20%5Bstyle%3Dradial%5D%0A%20%20%20%20%20%20%20%20%22P1%20%28top%2C%20back%29%22%20%5Bcolor%3Dred%5D%0A%20%20%20%20%20%20%20%20%22P2%20%28top%2C%20front%29%22%20%5Bcolor%3Dorange%5D%0A%20%20%20%20%20%20%20%20label%3D%22Top%20shelf%22%0A%20%20%20%20%20%20%20%20color%3Dgreen%0A%20%20%20%20%7D%0A%20%20%20%20subgraph%20cluster_steps%20%7B%0A%20%20%20%20%20%20%20%20label%3D%22Steps%22%0A%20%20%20%20%20%20%20%20node%20%5Bstyle%3Dfilled%5D%3B%0A%20%20%20%20%20%221.%20Entferne%20Gruppe%20auf%20Position%20P5%22%20%5Bcolor%3Dgreenyellow%5D%0A%20%20%20%20%20%222.%20Entferne%20Gruppe%20auf%20Position%20P6%22%20%5Bcolor%3Dgreen%5D%0A%20%20%20%20%20%223.%20Schiebe%20Gruppe%20P5%20auf%20Position%20P6%22%20%5Bcolor%3Dgreenyellow%5D%0A%20%20%20%20%20%224.%20Schiebe%20Gruppe%20P4%20auf%20Position%20P5%22%20%5Bcolor%3Dyellow%5D%0A%20%20%20%20%20%225.%20Schiebe%20Gruppe%20P3%20auf%20Position%20P4%22%20%5Bcolor%3Dgold%5D%0A%20%20%20%20%20%226.%20Schiebe%20Gruppe%20P2%20auf%20Position%20P3%22%20%5Bcolor%3Dorange%5D%0A%20%20%20%20%20%227.%20Schiebe%20Gruppe%20P1%20auf%20Position%20P2%22%20%5Bcolor%3Dred%5D%0A%20%20%20%20%20%20%20%20node%20%5Bstyle%3Dfilled%5D%3B%0A%20%20%20%20%20%20%20%20%221.%20Entferne%20Gruppe%20auf%20Position%20P5%22%20%0A%20%20%20%20%20%20%20%20-%3E%20%222.%20Entferne%20Gruppe%20auf%20Position%20P6%22%0A%20%20%20%20%20%20%20%20-%3E%20%223.%20Schiebe%20Gruppe%20P5%20auf%20Position%20P6%22%0A%20%20%20%20%20%20%20%20-%3E%20%224.%20Schiebe%20Gruppe%20P4%20auf%20Position%20P5%22%0A%20%20%20%20%20%20%20%20-%3E%20%225.%20Schiebe%20Gruppe%20P3%20auf%20Position%20P4%22%0A%20%20%20%20%20%20%20%20-%3E%20%226.%20Schiebe%20Gruppe%20P2%20auf%20Position%20P3%22%0A%20%20%20%20%20%20%20%20-%3E%20%227.%20Schiebe%20Gruppe%20P1%20auf%20Position%20P2%22%0A%20%20%20%20%0A%20%20%20%20%20%20%20%20color%3Dblack%3B%0A%20%20%20%20%20%20%20%20hidden%3Dtrue%0A%20%20%20%20%7D%0A%20%20%20%20subgraph%20cluster_1%20%7B%0A%20%20%20%20%20%20%20%20label%20%3D%20%22Gruppen%22%3B%0A%20%20%20%20%20%20%20%20node%20%5Bstyle%3Dfilled%5D%3B%0A%20%20%20%20%20%20%20%20%22G1%22%20%5Bcolor%3Dred%5D%0A%20%20%20%20%20%20%20%20%22G2%22%20%5Bcolor%3Dorange%5D%0A%20%20%20%20%20%20%20%20%22G3%22%20%5Bcolor%3Dgold%5D%0A%20%20%20%20%20%20%20%20%22G4%22%20%5Bcolor%3Dyellow%5D%0A%20%20%20%20%20%20%20%20%22G5%22%20%5Bcolor%3Dgreenyellow%5D%0A%20%20%20%20%20%20%20%20%22G6%22%20%5Bcolor%3Dgreen%5D%0A%20%20%20%20%20%20%20%20%0A%20%20%20%20%20%20%20%20node%20%5Bstyle%3Dfilled%5D%0A%20%20%20%20%20%20%20%20%7B%22G1%22%0A%20%20%20%20%20%20%20%20-%3E%22G2%22%20%0A%20%20%20%20%20%20%20%20-%3E%22G3%22%0A%20%20%20%20%20%20%20%20-%3E%22G4%22%0A%20%20%20%20%20%20%20%20-%3E%22G5%22%0A%20%20%20%20%20%20%20%20-%3E%22G6%22%5Bstyle%3Dinvis%5D%20%7Brank%3D0.1%7D%20%7D%0A%20%20%20%20%20%20%20%20color%3Dblack%3B%0A%20%20%20%20%20%20%20%20hidden%3Dtrue%3B%0A%20%20%20%20%20%20%20%20%7D%0A%09%7D%0A%7D%0A)). 

However, when rendering to quarto via 

```cmd
quarto render index.qmd --to XXXX
```

and these are the results I get for html, pdf and docx outputs:

html:  
![grafik](https://github.com/quarto-dev/quarto-cli/assets/80548781/eed15517-45b2-4368-981b-e9abf503cca2)

pdf:
![grafik](https://github.com/quarto-dev/quarto-cli/assets/80548781/ef4b9a70-dc6d-4a3c-bfe0-9794d874d6cc)

docx:

![grafik](https://github.com/quarto-dev/quarto-cli/assets/80548781/acd710c0-1205-4d26-a357-176ff3bbecbd)

---

Notably for html, the diagram is obviously malformatted. The standalone string `lorem ipsum` below the diagram is also missing.

Sidenote: This also occures when compiling via the quarto-package for R.

---

What am I doing wrong; am I missing something obvious here? 

Any help is appreciated.

Thank you,  
Sincerely,  
~Gw

---

10.07.2023 16:06: updated to fix wrong copy-paste of `dot` to `{dot}`

**Comment by cscheid** [comment]
Can you make sure the .qmd you posted is actually the one you're using? Quarto doesn't recognize the syntax you shared:

````
```dot
...
```
````

This, for me, produces a code printing without a diagram (which is the expected behavior):

<img width="287" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/285675/09bf14de-bef2-452c-b0e4-057e17dc20c6">

Your document probably contains some other input. It's important for us to have access to precisely the document you've used.

**Reply by Gewerd-Strauss** [reply]
Whoops. Dunno how that happened, thank you for pointing it out. Doesn't resolve it, but yea that's a stupid mistake. Code in original post is fixed now.

**Comment by mcanouil** [comment]
~~You cannot really compare to the online version, because it is not exactly the same version of Mermaid.js and they do a lot of updates, so differences are expected.
Online is 10.2.4 while Quarto is currently at 10.2.0-rc.2.~~

This being said your example is not reproducible, you are missing `{}` around dot.

**Reply by mcanouil** [reply]
Ahaha, funny weird thing, the `1. ` in the label causes the issue, I believe it is parsed as markdown list:

<table>
<tr><th>Quarto document</th><th>HTML</th></tr>
<tr>
<td>

````qmd
```{dot}
digraph G {
  label="Steps"
  "1. Entferne Gruppe auf Position P5"
}
```
````

</td>
<td>
<img width="577" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/e2533848-b176-4696-90f9-433f95cb67c6">
</td>
</tr>
<tr>
<td>

````qmd
```{dot}
digraph G {
  label="Steps"
  "- Entferne Gruppe auf Position P5"
}
```
````

</td>
<td>
<img width="577" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/07989b0d-d7bf-4933-8519-d5134a0ea6c1">
</td>
</tr>
<tr>
<td>

````qmd
```{dot}
digraph G {
  label="Steps"
  "1 Entferne Gruppe auf Position P5"
}
```
````

</td>
<td>
<img width="577" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/36153755-182d-4847-87a8-90f10a286723">
</td>
</tr>
</table>

**Reply by Gewerd-Strauss** [reply]
This seems to be the cause. Or rather, using something that is very much not existing list syntax resolves this; in my case tested with e.g. `"1: Entferne Gruppe auf Position P5"`.

I guess I'll make a few test sets cuz I am curious what else it might fail upon. 

But I wonder why quarto doesn't parse it as non-markdown - or rather, why does it parse the contents of the codeblock as markdown in the first place?

The documentation on [diagrams](https://quarto.org/docs/authoring/diagrams.html) describes them as markdown-_inspired_. To me that can mean a whole lot of different things, but it doesn't read as `This can process markdown`; rather such that its language design is comparable to markdown in terms of simplicity.

**Reply by cscheid** [reply]
> The documentation on [diagrams](https://quarto.org/docs/authoring/diagrams.html) describes them as markdown-inspired.

Just FYI: In that documentation, we are simply describing `mermaid` in their own words.

**Reply by cderv** [reply]
@mcanouil I think you idenfied the issue. Note that the `.md` intermediate looks like this  (first example)
````markdown
---
title: "figure divs"
format: html
keep-md: true
---

:::{.cell}

:::{.cell-output-display}

:::{}

:::{}
<svg width="672" height="480" viewbox="0.00 0.00 301.50 68.80" xmlns="http://www.w3.org/2000/svg" xlink="http://www.w3.org/1999/xlink" style="; max-width: none; max-height: none">
<g id="graph0" class="graph" transform="scale(1 1) rotate(0) translate(4 64.8)">
<title>G</title>
<polygon fill="white" stroke="transparent" points="-4,4 -4,-64.8 297.5,-64.8 297.5,4 -4,4"></polygon>
<text text-anchor="middle" x="146.75" y="-8.2" font-family="Times,serif" font-size="14.00">Steps</text>
<!-- 1. Entferne Gruppe auf Position P5 -->
<g id="node1" class="node">
<title>1. Entferne Gruppe auf Position P5</title>
<ellipse fill="none" stroke="black" cx="146.75" cy="-42.8" rx="147" ry="18"></ellipse>
<text text-anchor="middle" x="146.75" y="-38.6" font-family="Times,serif" font-size="14.00">1. Entferne Gruppe auf Position P5</text>
</g>
</g>
</svg>
:::
:::
:::
:::
````

As we can see the `<svg>` code is not protected as Raw HTML using a ` ```{=html} ` - that is why some content is parsed as Markdown (probably because [`markdown_in_html_blocks` extension](https://pandoc.org/MANUAL.html#extension-markdown_in_html_blocks) is enabled by Default in Pandoc.)

@cscheid any reason to not protect the `.svg` for the Diagram processing ? If not, it could be an easy fix.

**Reply by cscheid** [reply]
No, not at all, we should definitely do that.

**Reply by mcanouil** [reply]
I'll open an issue then ;)

**Comment by mcanouil** [comment, marked as answer]
Closing in favour of https://github.com/quarto-dev/quarto-cli/issues/6163.
Thanks @Gewerd-Strauss for reporting this!
